Error during macro expansion

<details>
  <summary>Stacktrace</summary>

```
Error during macro expansion

java.lang.IllegalStateException: java.lang.IllegalStateException: Check failed.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at kotlin.streams.jdk8.StreamsKt.toList(Streams.kt:60)
	at org.rust.lang.core.macros.MacroExpansionTaskBase$submitExpansionTask$1.invoke(MacroExpansionTask.kt:149)
	at org.rust.lang.core.macros.MacroExpansionTaskBase$submitExpansionTask$1.invoke(MacroExpansionTask.kt:39)
	at org.rust.stdext.CompletableFutureKt$supplyAsync$1.get(CompletableFuture.kt:18)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: Check failed.
	at org.rust.openapiext.UtilsKt.checkReadAccessNotAllowed(utils.kt:78)
	at org.rust.openapiext.UtilsKt.executeUnderProgressWithWriteActionPriorityWithRetries(utils.kt:241)
	at org.rust.lang.core.macros.MacroExpansionTaskBase$submitExpansionTask$1$stages2$2.apply(MacroExpansionTask.kt:136)
	at org.rust.lang.core.macros.MacroExpansionTaskBase$submitExpansionTask$1$stages2$2.apply(MacroExpansionTask.kt:39)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(Unknown Source)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
	at java.base/java.util.stream.AbstractTask.compute(Unknown Source)
	at java.base/java.util.concurrent.CountedCompleter.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.awaitJoin(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.get(Unknown Source)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:70)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:58)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.visitRestElementsAndCleanup(LocalInspectionsPass.java:331)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.inspect(LocalInspectionsPass.java:201)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.collectInformationWithProgress(LocalInspectionsPass.java:117)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:84)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:69)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$null$1(PassExecutorService.java:429)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1157)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:422)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:586)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:86)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:421)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:397)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:164)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:231)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:395)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:173)
	... 5 more
```

</details>

--- a/src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt
+++ b/src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt
@@ -54,10 +54,7 @@ import org.rust.stdext.cleanDirectory
 import org.rust.stdext.waitForWithCheckCanceled
 import java.nio.file.Path
 import java.nio.file.Paths
-import java.util.concurrent.Callable
-import java.util.concurrent.CompletableFuture
-import java.util.concurrent.CountDownLatch
-import java.util.concurrent.Future
+import java.util.concurrent.*
 
 interface MacroExpansionManager {
     val indexableDirectory: VirtualFile?
@@ -247,6 +244,18 @@ private class MacroExpansionServiceImplInner(
     }
 
     private val taskQueue = MacroExpansionTaskQueue(project)
+
+    /**
+     * We must use a separate pool because:
+     * 1. [ForkJoinPool.commonPool] is heavily used by the platform
+     * 2. [ForkJoinPool] can start execute a task when joining ([ForkJoinTask.get]) another task
+     * 3. the platform sometimes join ForkJoinTasks under read lock
+     * 4. for macro expansion it's critically important that tasks are executed without read lock.
+     *
+     * In short, use of [ForkJoinPool.commonPool] in this place leads to crashes.
+     * See [issue](https://github.com/intellij-rust/intellij-rust/issues/3966)
+     */
+    private val pool = Executors.newWorkStealingPool()
 
     private val dataFile: Path
         get() = dirs.dataFile
@@ -520,6 +529,7 @@ private class MacroExpansionServiceImplInner(
         class ProcessUnprocessedMacrosTask : MacroExpansionTaskBase(
             project,
             storage,
+            pool,
             dirs.expansionsDirVi.pathAsPath
         ) {
             override fun getMacrosToExpand(): Sequence<List<Extractable>> {
@@ -553,6 +563,7 @@ private class MacroExpansionServiceImplInner(
         class ProcessModifiedMacrosTask(private val workspaceOnly: Boolean) : MacroExpansionTaskBase(
             project,
             storage,
+            pool,
             dirs.expansionsDirVi.pathAsPath
         ) {
             override fun getMacrosToExpand(): Sequence<List<Extractable>> {

--- a/src/main/kotlin/org/rust/lang/core/macros/MacroExpansionTask.kt
+++ b/src/main/kotlin/org/rust/lang/core/macros/MacroExpansionTask.kt
@@ -39,6 +39,7 @@ import kotlin.system.measureNanoTime
 abstract class MacroExpansionTaskBase(
     project: Project,
     private val storage: ExpandedMacroStorage,
+    private val pool: Executor,
     private val realFsExpansionContentRoot: Path
 ) : Task.Backgroundable(project, "Expanding Rust macros", /* canBeCancelled = */ false) {
     private val transactionExecutor = TransactionExecutor(project)
@@ -118,7 +119,7 @@ abstract class MacroExpansionTaskBase(
         estimateStages.set(extractableList.size)
         doneStages.set(0)
 
-        val pool = ForkJoinPool.commonPool()
+        // All subsequent parallelStream tasks are executed on the same [pool]
         supplyAsync(pool) {
             realTaskIndicator.text2 = "Expanding macros"
 

